Fix validating B2C issuer with tfp in issuer URI.

pmaytak opened the pull request:
Issue #274 

Fix to validate the B2C issuer with `tfp` URI type (ex. `{domain}/{tfp}/{tid}/{userFlow}/v2.0/`) correctly.

More detail:
[B2C tokens compatibility](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tokens-overview#compatibility)
[Configure B2C JWT token compatibility](https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-tokens#configure-jwt-token-compatibility)

jennyf19 commented:
I agree, i'll ping b2c and see what they say about this. we don't want to necessarily support legacy stuff if we don't have to. i would like to get more context on that before moving forward.

pmaytak commented:
So, a couple of options:
1. Keep code as is - B2C tfp issuers with fail with generic invalid issuer exception.
2. Merge this PR - B2C tfp issuers will validate correctly.
3. At the end of the validate method, we can parse actual issuer URI, if it has tfp, then throw a more descriptive exception. 
https://github.com/AzureAD/microsoft-identity-web/blob/b694b720190fc2938ce58700b999ef35bf2ffd03/src/Microsoft.Identity.Web/Resource/AadIssuerValidator.cs#L128-L132

jmprieur commented:
@pmaytak. Given the reply from the B2C team and the customer, I think we could do 3.
- When you write parse, do you mean just verify that there is the /tpf/ string in it?
- We'd throw an exception explaining that this is not supported with an aka.ms link https://aka.ms/ms-id-web/b2c-issuer

pmaytak commented on B2C-Tfp-Issuer.md:
```diff
@@ -0,0 +1,14 @@
+# Azure AD B2C issuer claim
```
Better title?

jennyf19 replied:
i agree, i would put this in the wiki.

pmaytak replied:
Yes, it will be in the wiki. Added here just to get comments.

pmaytak commented:
@jmprieur @jennyf19 
Updated PR:
- throwing an exception with an aka.ms link.
- added B2C-Tfp-Issuer.md just to get feedback for the wiki (will not merge it)

jennyf19 commented on B2C-Tfp-Issuer.md:
```diff
@@ -0,0 +1,14 @@
+# Azure AD B2C issuer claim
+
+Azure AD B2C user flows can be customized to emit issuer claim in two formats: `https://<domain>/<tenant>/v2.0/` and `https://<domain>/tfp/<tenant>/<userFlow>/v2.0/`. Originally `tfp` (which stands for Trust Framework Policy) in the URI was used to differentiate between Azure AD and Azure AD B2C . Currently the B2C team is working to streamline the experience and remove `tfp`. Therefore, the curr ...
```
i don't think you need to explain this. Just mention we support the default value.

pmaytak replied:
@jennyf19 
I updated this wiki entry.

jennyf19 commented on B2C-Tfp-Issuer.md:
```diff
@@ -0,0 +1,14 @@
+
+Azure AD B2C user flows can be customized to emit issuer claim in two formats: `https://<domain>/<tenant>/v2.0/` and `https://<domain>/tfp/<tenant>/<userFlow>/v2.0/`. Originally `tfp` (which stands for Trust Framework Policy) in the URI was used to differentiate between Azure AD and Azure AD B2C . Currently the B2C team is working to streamline the experience and remove `tfp`. Therefore, the curr ...
+
+### Creating Azure B2C user flow
```
i'm not sure why we need this in our wiki, as it's in the b2c docs. plus user flow and policy are actually different, but we are using the term user flow to cover both, so i think this could lead to more confusion.

pmaytak replied:
I'll remove this section.

I thought user flow is just a predefined policy, right?

jennyf19 replied:
you can have [custom policies](https://docs.microsoft.com/en-us/azure/active-directory-b2c/custom-policy-overview#comparing-user-flows-and-custom-policies), so we don't distinguish between the two, but b2c does.

jennyf19 commented on src/Microsoft.Identity.Web/Constants/IDWebErrorMessage.cs:
```diff
@@ -32,6 +32,7 @@ internal static class IDWebErrorMessage
         public const string IssuerMetadataUrlIsRequired = "IDW10301: Azure AD Issuer metadata address URL is required. ";
         public const string NoMetadataDocumentRetrieverProvided = "IDW10302: No metadata document retriever is provided. ";
         public const string IssuerDoesNotMatchValidIssuers = "IDW10303: Issuer: '{0}', does not match any of the valid issuers provided for this application. ";
+        public const string B2CTfpIssuerNotSupported = "IDW10304: B2C issuer with 'tfp' in the URI is not supported. See https://aka.ms/ms-id-web/b2c-issuer for details.";
```
maybe something like "microsoft identity web does not support a B2C issuer with...."

jennyf19 commented:
let's not include the .md file for this. thx.
